### PR TITLE
Support --no-loc to skip location info in Json

### PR DESCRIPTION
**Pull request by Shaikh-Ubaid**
This PR supports a `--no-loc` flag which enables the user to skip the location information in the generated AST/R Json format.

```console
$ lfortran examples/expr2.f90 --show-asr --json --no-loc
{
    "node": "TranslationUnit",
    "fields": {
        "symtab": {
            "node": "SymbolTable1",
            "fields": {
                "expr2": {
                    "node": "Program",
                    "fields": {
                        "symtab": {
                            "node": "SymbolTable2",
                            "fields": {
                                "x": {
                                    "node": "Variable",
                                    "fields": {
                                        "parent_symtab": 2,
                                        "name": "x",
                                        "dependencies": [],
                                        "intent": "Local",
                                        "symbolic_value": [],
                                        "value": [],
                                        "storage": "Default",
                                        "type": {
                                            "node": "Integer",
                                            "fields": {
                                                "kind": 4
                                            }
                                        },
                                        "type_declaration": [],
                                        "abi": "Source",
                                        "access": "Public",
                                        "presence": "Required",
                                        "value_attr": false
                                    }
                                }
                            }
                        },
                        "name": "expr2",
                        "dependencies": [],
                        "body": [
                            {
                                "node": "Assignment",
                                "fields": {
                                    "target": {
                                        "node": "Var",
                                        "fields": {
                                            "v": "x (SymbolTable2)"
                                        }
                                    },
                                    "value": {
                                        "node": "IntegerBinOp",
                                        "fields": {
                                            "left": {
                                                "node": "IntegerBinOp",
                                                "fields": {
                                                    "left": {
                                                        "node": "IntegerConstant",
                                                        "fields": {
                                                            "n": 2,
                                                            "type": {
                                                                "node": "Integer",
                                                                "fields": {
                                                                    "kind": 4
                                                                }
                                                            }
                                                        }
                                                    },
                                                    "op": "Add",
                                                    "right": {
                                                        "node": "IntegerConstant",
                                                        "fields": {
                                                            "n": 3,
                                                            "type": {
                                                                "node": "Integer",
                                                                "fields": {
                                                                    "kind": 4
                                                                }
                                                            }
                                                        }
                                                    },
                                                    "type": {
                                                        "node": "Integer",
                                                        "fields": {
                                                            "kind": 4
                                                        }
                                                    },
                                                    "value": {
                                                        "node": "IntegerConstant",
                                                        "fields": {
                                                            "n": 5,
                                                            "type": {
                                                                "node": "Integer",
                                                                "fields": {
                                                                    "kind": 4
                                                                }
                                                            }
                                                        }
                                                    }
                                                }
                                            },
                                            "op": "Mul",
                                            "right": {
                                                "node": "IntegerConstant",
                                                "fields": {
                                                    "n": 5,
                                                    "type": {
                                                        "node": "Integer",
                                                        "fields": {
                                                            "kind": 4
                                                        }
                                                    }
                                                }
                                            },
                                            "type": {
                                                "node": "Integer",
                                                "fields": {
                                                    "kind": 4
                                                }
                                            },
                                            "value": {
                                                "node": "IntegerConstant",
                                                "fields": {
                                                    "n": 25,
                                                    "type": {
                                                        "node": "Integer",
                                                        "fields": {
                                                            "kind": 4
                                                        }
                                                    }
                                                }
                                            }
                                        }
                                    },
                                    "overloaded": []
                                }
                            },
                            {
                                "node": "Print",
                                "fields": {
                                    "values": [
                                        {
                                            "node": "Var",
                                            "fields": {
                                                "v": "x (SymbolTable2)"
                                            }
                                        }
                                    ],
                                    "separator": [],
                                    "end": []
                                }
                            }
                        ]
                    }
                }
            }
        },
        "items": []
    }
}
```

### CODE REVIEW

**Comment by Shaikh-Ubaid** [comment]
We have the flag --dump-all-passes, but it only prints the asr in text format. I often use fortran, the json and the visualisation format to debug codes. This PR supports applying all passes to fortran, json, viz and tree formats.

File: src/bin/lfortran.cpp [hunk 1]
```diff
@@ -2058,6 +2063,7 @@ int main(int argc, char *argv[])
         app.add_flag("--legacy-array-sections", compiler_options.legacy_array_sections, "Enables passing array items as sections if required");
         app.add_flag("--ignore-pragma", compiler_options.ignore_pragma, "Ignores all the pragmas");
         app.add_flag("--stack-arrays", compiler_options.stack_arrays, "Allocate memory for arrays on stack");
+        app.add_flag("--all-passes", compiler_options.all_passes, "Apply all ASR passes to --show-asr and --show-fortran");
```
Comment (certik) [inline src/bin/lfortran.cpp:2066]: Why not adding an option like `--dump-fortran` (or something like that) to dump Fortran together with Clojure in `--dump-all-passes`.

**Comment by certik** [comment]
How about this:

* `--show-asr` is telling LFortran to print ASR as Clojure like output, which is exactly 1:1 with the internal ASR representation in memory
* `--show-fortran` is telling LFortran to print ASR as Fortran code (directly)
* `--json` will print ASR as JSON
* `--tree`
* `--visualize`

By default the above prints ASR before any passes are applied.

Then `--dump-all-passes` will print the above for all passes.

`--last-pass` will print the ASR after the last pass.

**Comment by Thirumalai-Shaktivel** [comment]
@Shaikh-Ubaid I thought of working on `--dump-fortran`; did you start working on it?

**Comment by Shaikh-Ubaid** [comment]
>@Shaikh-Ubaid I thought of working on --dump-fortran; did you start working on it?

I will take care of these Thirumalai. Please focus on other tasks.

**Comment by Shaikh-Ubaid** [comment]
I updated the PR description. This is ready. Please let me know if a reference test is to be added. I feel it is not needed. (Or can be added later if needed).

File: src/libasr/pickle.h [hunk 2]
```diff
@@ -17,8 +17,8 @@ namespace LCompilers {
     std::string pickle_tree(ASR::TranslationUnit_t &asr, bool colors, bool show_intrinsic_modules=false);
 
     // Print Json structure
-    std::string pickle_json(ASR::asr_t &asr, LocationManager &lm, bool show_intrinsic_modules=false);
-    std::string pickle_json(ASR::TranslationUnit_t &asr, LocationManager &lm, bool show_intrinsic_modules=false);
+    std::string pickle_json(ASR::asr_t &asr, LocationManager &lm, bool no_loc=false, bool show_intrinsic_modules=false);
+    std::string pickle_json(ASR::TranslationUnit_t &asr, LocationManager &lm, bool no_loc=false, bool show_intrinsic_modules=false);
```
Comment (certik) [inline src/libasr/pickle.h:21]: Let's not use default arguments, it leads to issues with overloads and clarify of what the values actually are. For example, if there is any code in LFortran that just called `pickle_json(asr, lm, true)`, then after this PR this very same code now will evaluate to `no_loc=true` and `show_intrinsic_modules=false`, which is not what was intended.

To fix this, do it like this:
```suggestion
    std::string pickle_json(ASR::TranslationUnit_t &asr, LocationManager &lm, bool no_loc, bool show_intrinsic_modules);
```
And then modify all call sites to supply all arguments explicitly.

File: src/lfortran/pickle.h [hunk 3]
```diff
@@ -20,8 +20,8 @@ namespace LCompilers::LFortran {
     std::string pickle_tree(AST::TranslationUnit_t &ast, bool colors=true);
 
     // Print Json structure
-    std::string pickle_json(AST::ast_t &ast, LocationManager &lm);
-    std::string pickle_json(AST::TranslationUnit_t &ast, LocationManager &lm);
+    std::string pickle_json(AST::ast_t &ast, LocationManager &lm, bool no_loc=false);
+    std::string pickle_json(AST::TranslationUnit_t &ast, LocationManager &lm, bool no_loc=false);
```
Comment (certik) [inline src/lfortran/pickle.h:24]: same here, no default arguments.

**Comment by certik** [comment]
Otherwise it looks good.

**Review by certik — APPROVED** [review]

**Pull request merged**